### ISSUE TITLE
fish code in the background

### ISSUE THREAD

**Issue by JanKanis** [fish-shell/fish-shell#563]
Although I find fish much better than bash for everyday use, there's one major feature bash supports but fish does not: running fish code while doing anything else. Specifically, backgrounding a fish function doesn't work, and running fish code as part of a pipeline blocks the whole pipeline. 

Try this in bash:

```
$ function foo() {
> echo bar
> sleep 10
> }
$ foo | cat
bar
# notice that output arrives immediately, though the whole job only finishes after 10 seconds
$ foo &
[1] 1126
bar
$ # you can continue your interactive session. After 10 seconds: 
[1]+ Done        foo
```

If you try the same in fish, you'll notice that in a pipeline the output only arrives after the foo function has exited, i.e. after 10 seconds, and running a fish function in the background basically has no effect. I run into this sometimes when building pipelines to filter stuff or when using aliases, for example this one:

```
function gitg
    # to see a nice graphical history of the git repository without gtk error messages
    command gitg $argv ^/dev/null
end
```

If I then run `gitg &`, the shell is blocked until I quit gitg. 

Pressing ctrl-Z while a function or loop or other compound statement is being executed backgrounds only the current external command, and continues the fish code in the foreground:

```
> begin
      sleep 20
      echo done
  end
# press ^Z
Job 1, “sleep 20” has stopped
done
> bg
Send job 1 “sleep 20” to background
# some seconds later...
Job 1, “sleep 20” has ended
> 
```

Ideally this should background the entire block. (But this part does not work in bash either) In fact, backgrounding a compound statement that consists only of functions and builtins doesn't work at all: 

```
> while test 1; echo -n o; end
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo....
# lots of o's. ^Z doesn't do anything (though ^C works)
```

Fixing all of this would require some way of having fish execute multiple pieces of fish shellcode in parallel, so that is probably not a simple fix. However I think this issue should be discussed so we can decide on how to move forward with respect to fixing this at some point in the future. Fish's parser could really use an overhaul (see also bug #557), and if/when that happens this issue should also be taken into consideration.

**Comment by faho** [comment]
The backgrounding half of this is #238, the pipe-blocking is related to #936.

**Comment by floam** [comment]
> openat doesn't appear to be implemented on OS X. I see it in Linux.

FWIW not the case anymore.

**Comment by floam** [comment]
from the manpage:

```
HISTORY
     An open() function call appeared in Version 6 AT&T UNIX.  The openat()
     function was introduced in OS X 10.10
```

**Comment by floam** [comment]
(`symlinkat` also added)

**Comment by floam** [comment]
full list of added sys calls in OS X 10.10. I can't find any other new ones since 10.6.

 * `fchmodat`
 * `fchownat`
 * `getattrlistat`
 * `mkdirat`
 * `openat` 
 * `readlinkat`
 * `fstatat`
 * `symlinkat`
 * `unlinkat`

**Comment by floam** [comment]
full list of added *documented* syscalls in OS X 10.10:

 * `fchmodat`
 * `fchownat`
 * `getattrlistat`
 * `mkdirat`
 * `openat` 
 * `readlinkat`
 * `fstatat`
 * `symlinkat`
 * `unlinkat`

it would seem these are the first new public syscalls added since 10.5, which added `sendfile` and the now-deprecated `*stat64` and `*statfs64` syscalls.

**Comment by faho** [comment]
I'm closing this as a duplicate of #238.
